File: Mac/ES_R_DevKit_Mac/MEME_Academic/MEME_Academic/ChartService.swift
//
//  ChartService.swift
//  MEME_Academic
//
//  リアルタイムチャート用データ計算サービス。
//  受信した AcademicData をバッファリングし、
//  ChartPlot (yMin/yMax, series, 軸ラベル) を更新する。
//

import Foundation
import SwiftUI

@MainActor
final class ChartService {

    private var chartDatas: [AcademicData] = []

    /// 既存ロジックと同じ「直近 chartLimit 件」を取り出すための上限。
    /// xMax(200) - xLongScale(25)/2 = 187
    private let chartLimit: Int = 200 - 25 / 2

    // MARK: - Buffer

    func append(_ data: AcademicData) {
        chartDatas.append(data)
    }

    func reset() {
        chartDatas.removeAll()
    }

    var count: Int { chartDatas.count }

    // MARK: - Plot update

    func updatePlots(chart1: inout ChartPlot,
                     chart1Category: Int,
                     chart1Eog: EogToggles,
                     chart1Gyro: GyroToggles,
                     chart1Accel: AccelToggles,
                     chart2: inout ChartPlot,
                     chart2Category: Int,
                     chart2Eog: EogToggles,
                     chart2Gyro: GyroToggles,
                     chart2Accel: AccelToggles,
                     chart3: inout ChartPlot,
                     chart3Category: Int,
                     chart3Eog: EogToggles,
                     chart3Gyro: GyroToggles,
                     chart3Accel: AccelToggles) {
        let sliced = slicedDatas()
        let (xInit, xLabel) = computeXOffset(datasCount: chartDatas.count)

        updateChartPlot(&chart1,
                        category: chart1Category,
                        eog: chart1Eog, gyro: chart1Gyro, accel: chart1Accel,
                        sliced: sliced, xInitial: xInit, xLabel: xLabel)

        updateChartPlot(&chart2,
                        category: chart2Category,
                        eog: chart2Eog, gyro: chart2Gyro, accel: chart2Accel,
                        sliced: sliced, xInitial: xInit, xLabel: xLabel)

        updateChartPlot(&chart3,
                        category: chart3Category,
                        eog: chart3Eog, gyro: chart3Gyro, accel: chart3Accel,
                        sliced: sliced, xInitial: xInit, xLabel: xLabel)
    }

    // MARK: - Private helpers

    private func slicedDatas() -> [AcademicData] {
        let datas = chartDatas
        let limit = chartLimit
        if datas.count >= limit {
            let startIndex = datas.count - limit
            return Array(datas[startIndex..<datas.count])
        } else {
            return datas
        }
    }

    private func computeXOffset(datasCount: Int) -> (xInitial: Float, xLabel: Int) {
        let xMax: Float = 200
        let xLong: Int = 25
        if Float(datasCount) > xMax - Float(xLong) {
            let pos = datasCount % xLong
            let xInitial = Float(pos) * -1
            let xLabel = (datasCount / xLong) - (Int(xMax) / xLong) + 1
            return (xInitial, xLabel)
        }
        return (0, 0)
    }

    private func updateChartPlot(_ plot: inout ChartPlot,
                                 category: Int,
                                 eog: EogToggles,
                                 gyro: GyroToggles,
                                 accel: AccelToggles,
                                 sliced: [AcademicData],
                                 xInitial: Float,
                                 xLabel: Int) {
        plot.xInitial = xInitial
        plot.xLabelStart = xLabel

        switch category {
        case 0: // EOG
            plot.yMin = -1200; plot.yMax = 1200
            plot.series = buildEogSeries(sliced: sliced, toggles: eog)
        case 1: // Gyro (Full only)
            plot.yMin = -36000; plot.yMax = 36000
            plot.series = buildGyroSeries(sliced: sliced, toggles: gyro)
        case 2: // Accel
            plot.yMin = -36000; plot.yMax = 36000
            plot.series = buildAccelSeries(sliced: sliced, toggles: accel)
        default: break
        }
    }

    private func buildEogSeries(sliced: [AcademicData], toggles: EogToggles) -> [ChartSeries] {
        var L: [Double] = []
        var R: [Double] = []
        var H: [Double] = []
        var V: [Double] = []
        for d in sliced {
            if let s = d as? AcademicStandardData {
                if toggles.left { L.append(Double(s.eogL1)) }
                if toggles.right { R.append(Double(s.eogR1)) }
                if toggles.deltaH { H.append(Double(s.eogH1)) }
                if toggles.deltaV { V.append(Double(s.eogV1)) }
            } else if let f = d as? AcademicFullData {
                if toggles.left { L.append(Double(f.eogL)) }
                if toggles.right { R.append(Double(f.eogR)) }
                if toggles.deltaH { H.append(Double(f.eogH)) }
                if toggles.deltaV { V.append(Double(f.eogV)) }
            }
        }
        var out: [ChartSeries] = []
        if toggles.left { out.append(ChartSeries(name: "EOG L", color: .yellow, values: L)) }
        if toggles.right { out.append(ChartSeries(name: "EOG R", color: .green, values: R)) }
        if toggles.deltaH { out.append(ChartSeries(name: "ΔH", color: .red, values: H)) }
        if toggles.deltaV { out.append(ChartSeries(name: "ΔV", color: .blue, values: V)) }
        return out
    }

    private func buildGyroSeries(sliced: [AcademicData], toggles: GyroToggles) -> [ChartSeries] {
        var X: [Double] = []; var Y: [Double] = []; var Z: [Double] = []
        for d in sliced {
            guard let f = d as? AcademicFullData else { continue }
            if toggles.x { X.append(Double(f.gyroX)) }
            if toggles.y { Y.append(Double(f.gyroY)) }
            if toggles.z { Z.append(Double(f.gyroZ)) }
        }
        var out: [ChartSeries] = []
        if toggles.x { out.append(ChartSeries(name: "Gyro X", color: .red, values: X)) }
        if toggles.y { out.append(ChartSeries(name: "Gyro Y", color: .green, values: Y)) }
        if toggles.z { out.append(ChartSeries(name: "Gyro Z", color: .blue, values: Z)) }
        return out
    }

    private func buildAccelSeries(sliced: [AcademicData], toggles: AccelToggles) -> [ChartSeries] {
        var X: [Double] = []; var Y: [Double] = []; var Z: [Double] = []
        for d in sliced {
            if let s = d as? AcademicStandardData {
                if toggles.x { X.append(Double(s.accX) + UserSetting.getXAxis()) }
                if toggles.y { Y.append(Double(s.accY) + UserSetting.getYAxis()) }
                if toggles.z { Z.append(Double(s.accZ) + UserSetting.getZAxis()) }
            } else if let f = d as? AcademicFullData {
                if toggles.x { X.append(Double(f.accX) + UserSetting.getXAxis()) }
                if toggles.y { Y.append(Double(f.accY) + UserSetting.getYAxis()) }
                if toggles.z { Z.append(Double(f.accZ) + UserSetting.getZAxis()) }
            }
        }
        var out: [ChartSeries] = []
        if toggles.x { out.append(ChartSeries(name: "Acc X", color: .red, values: X)) }
        if toggles.y { out.append(ChartSeries(name: "Acc Y", color: .green, values: Y)) }
        if toggles.z { out.append(ChartSeries(name: "Acc Z", color: .blue, values: Z)) }
        return out
    }
}

File: windows/ES_R_DevKit_Windows/MEME_Academic_Sample/Services/ChartService.cs
using MEMELib_Academic;
using MEME_Academic_Sample.Charting;

namespace MEME_Academic_Sample.Services;

/// <summary>
/// リアルタイム／リプレイ チャート用のデータ計算サービス。
/// 受信したサンプルをフルレート(間引きなし)でバッファし、表示ウィンドウぶんの
/// <see cref="ChartPlot"/> を更新する。間引かないのはハム(50/60Hz)成分を波形に残すため。
/// Mac 版 ChartService の移植。
/// </summary>
public sealed class ChartService
{
    /// <summary>保持する最大サンプル数。最大ウィンドウ(30秒 × 100Hz)を賄えれば足りる。</summary>
    private const int MaxBufferSamples = 30 * 100;

    private readonly List<AcademicData> _samples = new(MaxBufferSamples);

    /// <summary>Append は BLE 受信スレッド、UpdatePlots は UI スレッドから呼ばれる。</summary>
    private readonly Lock _gate = new();

    /// <summary>_samples[0] のストリーム全体での絶対サンプル位置。</summary>
    private int _baseIndex;

    /// <summary>加速度に加算する Setting のオフセット。</summary>
    public double AccelOffsetX { get; set; }

    public double AccelOffsetY { get; set; }

    public double AccelOffsetZ { get; set; }

    /// <summary>X 軸のタイムスタンプをローカルタイムで表示するか(Setting 由来)。</summary>
    public bool ConvertToLocalTime { get; set; } = true;

    public int Count
    {
        get
        {
            lock (_gate)
            {
                return _samples.Count;
            }
        }
    }

    public void Append(AcademicData data)
    {
        lock (_gate)
        {
            _samples.Add(data);
            var overflow = _samples.Count - MaxBufferSamples;
            if (overflow > 0)
            {
                _samples.RemoveRange(0, overflow);
                _baseIndex += overflow;
            }
        }
    }

    /// <summary>
    /// バッファをクリアする。<paramref name="baseIndex"/> には次に Append される
    /// サンプルの絶対位置(リプレイのシーク先など)を渡す。通常は 0。
    /// </summary>
    public void Reset(int baseIndex = 0)
    {
        lock (_gate)
        {
            _samples.Clear();
            _baseIndex = baseIndex;
        }
    }

    /// <summary>各チャートの表示状態から <see cref="ChartPlot"/> を組み立て直す。</summary>
    public void UpdatePlots(
        IReadOnlyList<ChartPanel> panels,
        int sampleRate,
        int xRangeSeconds,
        IReadOnlyDictionary<int, string>? artifacts = null)
    {
        var rate = Math.Max(sampleRate, 1);
        var windowSamples = Math.Max(1, xRangeSeconds * rate);

        // 受信スレッドの Append と競合しないよう、表示ウィンドウをここで一度だけ取り出す。
        AcademicData[] window;
        int latestSampleIndex;
        DateTime? latestSampleUtc;
        lock (_gate)
        {
            // 直近 windowSamples 件のみ描画対象にする(波形は右詰めで、右端が最新サンプル)。
            var windowStart = Math.Max(0, _samples.Count - windowSamples);
            window = _samples.GetRange(windowStart, _samples.Count - windowStart).ToArray();
            latestSampleIndex = _baseIndex + Math.Max(_samples.Count - 1, 0);
            latestSampleUtc = _samples.Count > 0 ? _samples[^1].RecordedUtc : null;
        }

        var visibleArtifacts = ArtifactsInWindow(artifacts, latestSampleIndex, windowSamples);

        foreach (var panel in panels)
        {
            var plot = panel.Plot;
            plot.WindowSamples = windowSamples;
            plot.LatestSampleIndex = latestSampleIndex;
            plot.SampleRate = rate;
            plot.LatestSampleUtc = latestSampleUtc;
            plot.ConvertToLocalTime = ConvertToLocalTime;
            plot.Artifacts = visibleArtifacts;
            plot.Series = panel.AppliedCategory switch
            {
                ChartCategory.Electrooculography => BuildEogSeries(window, panel.Eog),
                ChartCategory.Gyroscope => BuildGyroSeries(window, panel.Gyro),
                ChartCategory.Accelerometer => BuildAccelSeries(window, panel.Accel),
                _ => [],
            };
        }
    }

    /// <summary>可視ウィンドウ(右端 latest、幅 windowSamples)に入る Artifact だけ抽出する。</summary>
    private static IReadOnlyList<ChartArtifact> ArtifactsInWindow(
        IReadOnlyDictionary<int, string>? artifacts, int latest, int windowSamples)
    {
        if (artifacts is null || artifacts.Count == 0)
        {
            return [];
        }

        var leftAbs = latest - windowSamples;
        var result = new List<ChartArtifact>();
        foreach (var (index, text) in artifacts)
        {
            if (index >= leftAbs && index <= latest)
            {
                result.Add(new ChartArtifact(index, text));
            }
        }

        return result;
    }

    /// <summary>
    /// EOG。Standard は 1 組目(EogL1 / EogR1 / EogH1 / EogV1)を、Full は EogL / EogR / EogH / EogV を描く。
    /// Quaternion モードには EOG が無いので何も出ない(Mac 版と同じ)。
    /// </summary>
    private static List<ChartSeries> BuildEogSeries(AcademicData[] window, EogToggles toggles)
    {
        var result = new List<ChartSeries>(4);
        if (toggles.Left)
        {
            result.Add(new ChartSeries("EOG L", ChartColors.EogLeft, Collect(window, d => d switch
            {
                AcademicStandardData s => s.EogL1,
                AcademicFullData f => f.EogL,
                _ => (double?)null,
            })));
        }

        if (toggles.Right)
        {
            result.Add(new ChartSeries("EOG R", ChartColors.EogRight, Collect(window, d => d switch
            {
                AcademicStandardData s => s.EogR1,
                AcademicFullData f => f.EogR,
                _ => (double?)null,
            })));
        }

        if (toggles.DeltaH)
        {
            result.Add(new ChartSeries("ΔH", ChartColors.EogDeltaH, Collect(window, d => d switch
            {
                AcademicStandardData s => s.EogH1,
                AcademicFullData f => f.EogH,
                _ => (double?)null,
            })));
        }

        if (toggles.DeltaV)
        {
            result.Add(new ChartSeries("ΔV", ChartColors.EogDeltaV, Collect(window, d => d switch
            {
                AcademicStandardData s => s.EogV1,
                AcademicFullData f => f.EogV,
                _ => (double?)null,
            })));
        }

        return result;
    }

    /// <summary>ジャイロは Full モードにしか無い。</summary>
    private static List<ChartSeries> BuildGyroSeries(AcademicData[] window, AxisToggles toggles)
    {
        var result = new List<ChartSeries>(3);
        if (toggles.X)
        {
            result.Add(new ChartSeries("Gyro X", ChartColors.AxisX,
                Collect(window, d => d is AcademicFullData f ? f.GyroX : null)));
        }

        if (toggles.Y)
        {
            result.Add(new ChartSeries("Gyro Y", ChartColors.AxisY,
                Collect(window, d => d is AcademicFullData f ? f.GyroY : null)));
        }

        if (toggles.Z)
        {
            result.Add(new ChartSeries("Gyro Z", ChartColors.AxisZ,
                Collect(window, d => d is AcademicFullData f ? f.GyroZ : null)));
        }

        return result;
    }

    private List<ChartSeries> BuildAccelSeries(AcademicData[] window, AxisToggles toggles)
    {
        var result = new List<ChartSeries>(3);
        if (toggles.X)
        {
            result.Add(new ChartSeries("Acc X", ChartColors.AxisX, Collect(window, d => d switch
            {
                AcademicStandardData s => s.AccX + AccelOffsetX,
                AcademicFullData f => f.AccX + AccelOffsetX,
                _ => (double?)null,
            })));
        }

        if (toggles.Y)
        {
            result.Add(new ChartSeries("Acc Y", ChartColors.AxisY, Collect(window, d => d switch
            {
                AcademicStandardData s => s.AccY + AccelOffsetY,
                AcademicFullData f => f.AccY + AccelOffsetY,
                _ => (double?)null,
            })));
        }

        if (toggles.Z)
        {
            result.Add(new ChartSeries("Acc Z", ChartColors.AxisZ, Collect(window, d => d switch
            {
                AcademicStandardData s => s.AccZ + AccelOffsetZ,
                AcademicFullData f => f.AccZ + AccelOffsetZ,
                _ => (double?)null,
            })));
        }

        return result;
    }

    /// <summary>そのモードに存在しない値(selector が null)は詰めずに飛ばす。</summary>
    private static double[] Collect(AcademicData[] window, Func<AcademicData, double?> selector)
    {
        var values = new List<double>(window.Length);
        foreach (var sample in window)
        {
            if (selector(sample) is { } value)
            {
                values.Add(value);
            }
        }

        return [.. values];
    }
}
